File: app/plugins/capacitor-databus-telemetry/ios/Sources/DatabusTelemetryPlugin/DatabusTelemetryPlugin.swift
import Foundation
import CoreLocation
import Capacitor

/**
 * Capacitor iOS plugin for the production telemetry hot path.
 *
 * Wires the TS `DatabusTelemetryPlugin` contract (`src/definitions.ts`) to the
 * native [DatabusTelemetry] engine: MQTT over TCP+TLS 8883 (CocoaMQTT),
 * CLLocationManager GPS with background-location mode, and a native
 * store-and-forward buffer.
 *
 * The plugin owns the single [DatabusTelemetry] instance and is the native→JS
 * event bridge: it sets closures on the engine that call `notifyListeners`.
 * `start/stop/checkPermissions/requestPermissions` are exported to JS via the
 * `pluginMethods` array (Capacitor's Objective-C runtime requires `@objc`).
 *
 * @since 0.0.1
 */
@objc(DatabusTelemetryPlugin)
public class DatabusTelemetryPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "DatabusTelemetryPlugin"
    public let jsName = "DatabusTelemetry"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "start", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stop", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermissions", returnType: CAPPluginReturnPromise),
    ]

    private let implementation = DatabusTelemetry()

    public override func load() {
        super.load()
        // Wire the native event stream to JS via notifyListeners.
        implementation.onStatus = { [weak self] status, message in
            self?.notifyListeners("status", data: [
                "status": status.rawValue,
                "message": message as Any,
            ])
        }
        implementation.onFix = { [weak self] fix in
            var data: [String: Any] = [
                "latitude": fix.latitude,
                "longitude": fix.longitude,
            ]
            if let b = fix.bearing { data["bearing"] = b }
            if let s = fix.speed { data["speed"] = s }
            if let t = fix.timestamp { data["timestamp"] = t }
            self?.notifyListeners("lastFix", data: data)
        }
        implementation.onQueuedCount = { [weak self] count in
            self?.notifyListeners("queuedCount", data: ["count": count])
        }
    }

    // MARK: - Plugin methods

    @objc func start(_ call: CAPPluginCall) {
        guard let vehicleId = call.getString("vehicleId"), !vehicleId.isEmpty else {
            call.reject("INVALID_PARAMETER: vehicleId is required")
            return
        }
        let host = brokerHostFromConfig()
        guard !host.isEmpty else {
            call.reject("plugins.DatabusTelemetry.brokerHost is not set in capacitor.config.ts")
            return
        }
        implementation.start(
            host: host,
            port: call.getInt("brokerPort"),
            useTls: call.getBool("useTls"),
            clientIdPrefix: clientIdPrefixFromConfig(),
            gpsIntervalMs: gpsIntervalMsFromConfig(),
            gpsMinDistanceM: gpsMinDistanceMFromConfig(),
            bufferMaxSize: bufferMaxSizeFromConfig(),
            vehicleId: vehicleId,
            username: call.getString("username"),
            token: call.getString("token"),
            caCertAsset: call.getString("caCertAsset")
        )
        call.resolve()
    }

    @objc func stop(_ call: CAPPluginCall) {
        implementation.stop()
        call.resolve()
    }

    @objc func checkPermissions(_ call: CAPPluginCall) {
        call.resolve(["location": locationPermissionState()])
    }

    @objc func requestPermissions(_ call: CAPPluginCall) {
        // CLLocationManager authorization is driven by the engine on start();
        // here we just report the current state. iOS does not expose a JS-side
        // permission prompt distinct from starting location updates.
        call.resolve(["location": locationPermissionState()])
    }

    // MARK: - Config accessors (Capacitor config: plugins.DatabusTelemetry)

    private func brokerHostFromConfig() -> String {
        getConfig().getString("brokerHost") ?? ""
    }

    private func clientIdPrefixFromConfig() -> String {
        getConfig().getString("clientIdPrefix") ?? "databus-"
    }

    private func gpsIntervalMsFromConfig() -> Double {
        let v = getConfig().getDouble("gpsIntervalMs", 5000.0)
        return max(v, 1000.0)
    }

    private func gpsMinDistanceMFromConfig() -> Double {
        getConfig().getDouble("gpsMinDistanceM", 5.0)
    }

    private func bufferMaxSizeFromConfig() -> Int {
        let v = getConfig().getInt("bufferMaxSize", 2000)
        return max(v, 1)
    }

    // MARK: - Permissions

    private func locationPermissionState() -> String {
        switch implementation.authorizationStatus {
        case .authorizedAlways, .authorizedWhenInUse:
            return "granted"
        case .denied, .restricted:
            return "denied"
        default:
            return "prompt"
        }
    }
}

File: app/plugins/capacitor-databus-telemetry/android/src/main/kotlin/org/simovi/databus/telemetry/DatabusTelemetryPlugin.kt
package org.simovi.databus.telemetry

import android.Manifest
import com.getcapacitor.JSObject
import com.getcapacitor.PermissionState
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.getcapacitor.annotation.PermissionCallback

/**
 * Capacitor Android plugin for the production telemetry hot path.
 *
 * Wires the TS `DatabusTelemetryPlugin` contract (`src/definitions.ts`) to the
 * native [TelemetryController] / [TelemetryService] / [TelemetryEngine]:
 * MQTT over TCP+TLS 8883, FusedLocation GPS, an Android foreground service
 * for background survival, and a native store-and-forward buffer.
 *
 * The plugin instance is the event bridge: it implements [PluginEmitter] and
 * registers itself with [TelemetryController] so native events reach JS via
 * `notifyListeners`. Permissions are handled through Capacitor's alias system
 * (foreground location alias; background location is a separate app-driven
 * prompt on Android 10+ — see README + `app/NATIVE-PERMISSIONS.md`).
 *
 * @since 0.0.1
 */
@CapacitorPlugin(
    name = "DatabusTelemetry",
    permissions = [
        Permission(
            alias = "location",
            strings = [
                Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.ACCESS_COARSE_LOCATION,
            ],
        ),
    ],
)
class DatabusTelemetryPlugin : Plugin(), PluginEmitter {

    override fun load() {
        super.load()
        // Register as the native→JS event bridge. Held for the plugin's life.
        TelemetryController.pluginEmitter = this
    }

    // --- plugin methods -----------------------------------------------------

    @PluginMethod
    fun start(call: PluginCall) {
        val vehicleId = call.getString("vehicleId")
        if (vehicleId.isNullOrEmpty()) {
            call.reject("INVALID_PARAMETER: vehicleId is required")
            return
        }
        if (getPermissionState("location") != PermissionState.GRANTED) {
            // Defer until the permission callback resolves.
            requestPermissionForAlias("location", call, "startPermissionCallback")
            return
        }
        beginStart(call, vehicleId)
    }

    @PermissionCallback
    private fun startPermissionCallback(call: PluginCall) {
        if (getPermissionState("location") == PermissionState.GRANTED) {
            val vehicleId = call.getString("vehicleId")
            if (vehicleId.isNullOrEmpty()) {
                call.reject("INVALID_PARAMETER: vehicleId is required")
                return
            }
            beginStart(call, vehicleId)
        } else {
            call.reject("PERMISSION_DENIED: location permission is required")
        }
    }

    private fun beginStart(call: PluginCall, vehicleId: String) {
        val config = readConfig()
        val options = TelemetryStartOptions(
            vehicleId = vehicleId,
            brokerHost = call.getString("brokerHost"),
            brokerPort = call.getInt("brokerPort"),
            useTls = call.getBoolean("useTls", config.useTls),
            username = call.getString("username"),
            token = call.getString("token"),
            caCertAsset = call.getString("caCertAsset"),
        )
        try {
            TelemetryController.start(context, config, options)
            call.resolve()
        } catch (e: Exception) {
            call.reject("Failed to start telemetry: ${e.message}")
        }
    }

    @PluginMethod
    fun stop(call: PluginCall) {
        try {
            TelemetryController.stop(context)
        } catch (e: Exception) {
            // Idempotent — swallow teardown errors, still resolve.
        }
        call.resolve()
    }

    @PluginMethod
    fun checkPermissions(call: PluginCall) {
        val state = getPermissionState("location")
        call.resolve(JSObject().put("location", stateToJs(state)))
    }

    @PluginMethod
    fun requestPermissions(call: PluginCall) {
        if (getPermissionState("location") != PermissionState.GRANTED) {
            requestPermissionForAlias("location", call, "requestPermissionCallback")
        } else {
            call.resolve(JSObject().put("location", "granted"))
        }
    }

    @PermissionCallback
    private fun requestPermissionCallback(call: PluginCall) {
        call.resolve(JSObject().put("location", stateToJs(getPermissionState("location"))))
    }

    // --- PluginEmitter (native → JS events) ---------------------------------

    override fun emitStatus(status: String, message: String?) {
        val data = JSObject()
        data.put("status", status)
        if (message != null) data.put("message", message)
        notifyListeners("status", data)
    }

    override fun emitFix(fix: TelemetryFix) {
        val data = JSObject()
        data.put("latitude", fix.latitude)
        data.put("longitude", fix.longitude)
        fix.bearing?.let { data.put("bearing", it) }
        fix.speed?.let { data.put("speed", it) }
        fix.timestamp?.let { data.put("timestamp", it) }
        notifyListeners("lastFix", data)
    }

    override fun emitQueuedCount(count: Int) {
        val data = JSObject()
        data.put("count", count)
        notifyListeners("queuedCount", data)
    }

    // --- helpers ------------------------------------------------------------

    private fun readConfig(): TelemetryConfig {
        val cfg = getConfig()
        val host = cfg.getString("brokerHost", "")
        if (host.isEmpty()) {
            // Surface a clear error rather than connecting to a nonsense host.
            throw IllegalStateException(
                "plugins.DatabusTelemetry.brokerHost is not set in capacitor.config.ts",
            )
        }
        return TelemetryConfig(
            brokerHost = host,
            brokerPort = cfg.getInt("brokerPort", TelemetryConfig.DEFAULT_BROKER_PORT),
            useTls = cfg.getBoolean("useTls", true),
            clientIdPrefix = cfg.getString("clientIdPrefix", TelemetryConfig.DEFAULT_CLIENT_ID_PREFIX)
                ?: TelemetryConfig.DEFAULT_CLIENT_ID_PREFIX,
            gpsIntervalMs = cfg.getDouble("gpsIntervalMs", TelemetryConfig.DEFAULT_GPS_INTERVAL_MS.toDouble())
                .toLong()
                .coerceAtLeast(1000L),
            gpsMinDistanceM = cfg.getDouble("gpsMinDistanceM", TelemetryConfig.DEFAULT_GPS_MIN_DISTANCE_M.toDouble())
                .toFloat(),
            bufferMaxSize = cfg.getInt("bufferMaxSize", TelemetryConfig.DEFAULT_BUFFER_MAX_SIZE)
                .coerceAtLeast(1),
            notificationChannelId = cfg.getString("notificationChannelId", TelemetryConfig.DEFAULT_NOTIFICATION_CHANNEL_ID)
                ?: TelemetryConfig.DEFAULT_NOTIFICATION_CHANNEL_ID,
            notificationTitle = cfg.getString("notificationTitle", TelemetryConfig.DEFAULT_NOTIFICATION_TITLE)
                ?: TelemetryConfig.DEFAULT_NOTIFICATION_TITLE,
        )
    }

    private fun stateToJs(state: PermissionState?): String = when (state) {
        PermissionState.GRANTED -> "granted"
        PermissionState.DENIED -> "denied"
        else -> "prompt"
    }
}
